Upgrade to 9.2.10.0 breaks project using asciidoctor-maven-plugin

**Environment Information**
  
Provide at least:
* JRuby version (`jruby -v`) and command line (flags, JRUBY_OPTS, etc): 9.2.10.0, maven
* Operating system and platform (e.g. `uname -a`): Linux

Other relevant info you may wish to add:

**Expected Behavior**

* Describe your expectation of how JRuby should behave, perhaps by showing how CRuby/MRI behaves.

Smooth upgrade from JRuby 9.2.9.0 to 9.2.10.0.

* Provide an executable Ruby script or a link to an example repository.

**Actual Behavior**

* Describe or show the actual behavior.

Not 100% sure if this is JRuby's fault, but project worked with 9.2.9.0, and only change being upgrade to 9.2.10.0 causes this:

```
[ERROR] Failed to execute goal org.asciidoctor:asciidoctor-maven-plugin:2.0.0-RC.1:process-asciidoc (output-pdf) on project: Execution output-pdf of goal org.asciidoctor:asciidoctor-maven-plugin:2.0.0-RC.1:process-asciidoc failed: (LoadError) no such file to load -- asciidoctor-pdf -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.asciidoctor:asciidoctor-maven-plugin:2.0.0-RC.1:process-asciidoc (output-pdf) on project: Execution output-pdf of goal org.asciidoctor:asciidoctor-maven-plugin:2.0.0-RC.1:process-asciidoc failed: (LoadError) no such file to load -- asciidoctor-pdf
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.apache.maven.wrapper.BootstrapMainStarter.start (BootstrapMainStarter.java:39)
    at org.apache.maven.wrapper.WrapperExecutor.execute (WrapperExecutor.java:122)
    at org.apache.maven.wrapper.MavenWrapperMain.main (MavenWrapperMain.java:61)
Caused by: org.apache.maven.plugin.PluginExecutionException: Execution output-pdf of goal org.asciidoctor:asciidoctor-maven-plugin:2.0.0-RC.1:process-asciidoc failed: (LoadError) no such file to load -- asciidoctor-pdf
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:148)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.apache.maven.wrapper.BootstrapMainStarter.start (BootstrapMainStarter.java:39)
    at org.apache.maven.wrapper.WrapperExecutor.execute (WrapperExecutor.java:122)
    at org.apache.maven.wrapper.MavenWrapperMain.main (MavenWrapperMain.java:61)
Caused by: org.jruby.exceptions.LoadError: (LoadError) no such file to load -- asciidoctor-pdf
    at org.jruby.RubyKernel.require (org/jruby/RubyKernel.java:974)
    at RUBY.require (uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/rubygems/core_ext/kernel_require.rb:54)
    at RUBY.<main> (<script>:1)
```
with plugin configured like this
```xml
            <plugin>
                <groupId>org.asciidoctor</groupId>
                <artifactId>asciidoctor-maven-plugin</artifactId>
                <version>2.0.0-RC.1</version><!-- using 1.6.0 doesn't help -->
                <executions>
                    <execution>
                        <id>output-pdf</id>
                        <phase>generate-resources</phase>
                        <goals>
                            <goal>process-asciidoc</goal>
                        </goals>
                        <configuration>
                            <backend>pdf</backend>
                        </configuration>
                    </execution>
                </executions>
                <dependencies>
                    <dependency>
                        <groupId>org.asciidoctor</groupId>
                        <artifactId>asciidoctorj-pdf</artifactId>
                        <version>1.5.0</version>
                    </dependency>
                    <dependency>
                        <groupId>org.jruby</groupId>
                        <artifactId>jruby-complete</artifactId>
                        <version>9.2.10.0</version>
                    </dependency>
                </dependencies>
            </plugin>
```

--- a/core/src/main/java/org/jruby/RubyDir.java
+++ b/core/src/main/java/org/jruby/RubyDir.java
@@ -239,20 +239,18 @@ public class RubyDir extends RubyObject implements Closeable {
             } else {
                 String[] keys = flags != null ? BASE_FLAGS : BASE;
                 IRubyObject[] rets = ArgsUtil.extractKeywordArgs(context, (RubyHash) tmp, keys);
-                String base = rets[0] == null || rets[0] == context.nil ? "" : RubyFile.get_path(context, rets[0]).asJavaString();
+                if (rets[0] == null || rets[0].isNil()) return "";
+                RubyString path = RubyFile.get_path(context, rets[0]);
+                Encoding[] enc = {path.getEncoding()};
+                String base = path.getUnicodeValue();
 
                 // Deep down in glob things are unhappy if base is not absolute.
                 if (!base.isEmpty()) {
-                    JRubyFile file = new JRubyFile(base);
-
-                    // FIXME: This is a bit of a hot mess.  Deep down in the ancient caves of glob land we do not
-                    // like see a base which is not an absolute path.  JRubyFile by itself uses what I am guessing
-                    // is the JVM CWD and not the runtimes CWD so if path is not absolute I remake the thing again...
-                    if (!file.isAbsolute()) {
-                        base = new JRubyFile(runtime.getCurrentDirectory(), base).getAbsolutePath();
-                    }
+                    base = RubyFile.expandPath(context, base, enc, runtime.getCurrentDirectory(), true, false);
                 }
+
                 if (flags != null) flags[0] = rets[1] == null ? 0 : RubyNumeric.num2int(rets[1]);
+
                 return base;
             }
         }
@@ -302,7 +300,7 @@ public class RubyDir extends RubyObject implements Closeable {
         String base = globOptions(context, args, flags);
         List<ByteList> dirs;
 
-        if (!(base == null || base.isEmpty()) && !(new File(base).exists())){
+        if (base != null && !base.isEmpty() && !(JRubyFile.createResource(context, base).exists())){
             dirs = new ArrayList<ByteList>();
         } else {
             IRubyObject tmp = args[0].checkArrayType();

--- a/core/src/main/java/org/jruby/RubyFile.java
+++ b/core/src/main/java/org/jruby/RubyFile.java
@@ -1717,9 +1717,17 @@ public class RubyFile extends RubyIO implements EncodingCapable {
     }
 
     static RubyString expandPathInternal(ThreadContext context, RubyString path, RubyString wd, boolean expandUser, boolean canonicalize) {
-        Ruby runtime = context.runtime;
-
         String relativePath = path.getUnicodeValue();
+        Encoding[] enc = {path.getEncoding()};
+        String cwd = wd == null ? null : wd.toString();
+        String expanded =  expandPath(context, relativePath, enc, cwd, expandUser, canonicalize);
+
+        return RubyString.newString(context.runtime, expanded, enc[0]);
+
+    }
+
+    public static String expandPath(ThreadContext context, String relativePath, Encoding[] enc, String cwd, boolean expandUser, boolean canonicalize) {
+        Ruby runtime = context.runtime;
 
         // Encoding logic lives in MRI's rb_file_expand_path_internal and should roughly equate to the following:
         // * Paths expanded from the system, like ~ expanding to the user's home dir, should be filesystem-encoded.
@@ -1731,13 +1739,13 @@ public class RubyFile extends RubyIO implements EncodingCapable {
         //
         // See dac9850 and jruby/jruby#3849.
 
-        Encoding enc = path.getEncoding();
         // for special paths like ~
         Encoding fsenc = runtime.getEncodingService().getFileSystemEncoding();
 
         // Special /dev/null of windows
         if (Platform.IS_WINDOWS && ("NUL:".equalsIgnoreCase(relativePath) || "NUL".equalsIgnoreCase(relativePath))) {
-            return RubyString.newString(runtime, concat("//./", relativePath.substring(0, 3)), fsenc);
+            enc[0] = fsenc;
+            return concat("//./", relativePath.substring(0, 3));
         }
 
         // treat uri-like and jar-like path as absolute
@@ -1804,21 +1812,21 @@ public class RubyFile extends RubyIO implements EncodingCapable {
                 }
             }
 
-            return concatStrings(runtime, preFix, extra, relativePath, enc);
+            return concatStrings(preFix, extra, relativePath);
         }
 
         String[] uriParts = splitURI(relativePath);
 
         // Handle ~user paths
         if (expandUser && startsWith(relativePath, '~')) {
-            enc = fsenc;
+            enc[0] = fsenc;
             relativePath = expandUserPath(context, relativePath, true);
         }
 
         if (uriParts != null) {
             //If the path was an absolute classpath path, return it as-is.
             if (uriParts[0].equals("classpath:")) {
-                return concatStrings(runtime, preFix, relativePath, postFix, enc);
+                return concatStrings(preFix, relativePath, postFix);
             }
             relativePath = uriParts[1];
         }
@@ -1826,11 +1834,9 @@ public class RubyFile extends RubyIO implements EncodingCapable {
         // Now that we're not treating it as a URI, we need to honor the canonicalize flag.
         // Do not insert early returns below.
 
-        String cwd;
         // If there's a second argument, it's the path to which the first argument is relative.
-        if (wd != null) {
-            cwd = wd.toString();
-            enc = fsenc;
+        if (cwd != null) {
+            enc[0] = fsenc;
             if (!cwd.startsWith("uri:")) {
                 // Handle ~user paths.
                 if (expandUser) {
@@ -1938,7 +1944,7 @@ public class RubyFile extends RubyIO implements EncodingCapable {
 
         if (postFix.contains("..")) postFix = adjustPostFixDotDot(postFix);
 
-        return concatStrings(runtime, preFix, realPath, postFix, enc);
+        return concatStrings(preFix, realPath, postFix);
     }
 
     private static String canonicalNormalized(CharSequence realPath) {
@@ -1961,10 +1967,8 @@ public class RubyFile extends RubyIO implements EncodingCapable {
         return postFix;
     }
 
-    private static RubyString concatStrings(final Ruby runtime, String s1, CharSequence s2, String s3, Encoding enc) {
-        StringBuilder str =
-                new StringBuilder(s1.length() + s2.length() + s3.length()).append(s1).append(s2).append(s3);
-        return new RubyString(runtime, runtime.getString(), str, enc);
+    private static String concatStrings(String s1, CharSequence s2, String s3) {
+        return new StringBuilder(s1.length() + s2.length() + s3.length()).append(s1).append(s2).append(s3).toString();
     }
 
     private static String canonicalizePath(String path) {

--- a/core/src/main/java/org/jruby/util/StringSupport.java
+++ b/core/src/main/java/org/jruby/util/StringSupport.java
@@ -172,6 +172,10 @@ public final class StringSupport {
         return new StringBuilder(str1.length() + str2.length()).append(str1).append(str2);
     }
 
+    public static String concat(final String str1, final String str2) {
+        return new StringBuilder(str1.length() + str2.length()).append(str1).append(str2).toString();
+    }
+
     public static String delete(final String str, final char c) { // str.replaceAll(c.toString(), "")
         char[] ary = null; int end = 0, s = 0;
         for (int i = 0; i < str.length(); i++) {

--- a/test/jruby.index
+++ b/test/jruby.index
@@ -95,8 +95,7 @@ jruby/test_uri_classloader
 jruby/test_rexml_document
 jruby/test_missing_jruby_home
 jruby/test_ast_inspector
-# Disabled temporarily, see jruby/jruby#6060
-#jruby/test_jarred_gems
+jruby/test_jarred_gems
 jruby/test_dir_with_plusses
 jruby/test_jar
 jruby/test_require

